### ISSUE TITLE
Use with Arm MPS2 development board?

### ISSUE THREAD

**Issue by daldr-ntml** [ARM-software/vscode-embedded-debug#17]
Hi, can you please offer me any guidance on how to use this extension to debug a Cortex M4 elf file, built using armclang, on an Arm MPS2 board, using the CMSIS-DAP USB interface?

**Comment by thegecko** [comment]
Hi @DavidA2014 

You need to outline a `launch.json` file of the form

```jsonc
{
            "name": "Debug",
            "type": "embedded-debug",
            "request": "launch",
            "serialNumber": "<usb device serial number>",
            "cmsisPack": "<path to a cmsis pack or reference in the form VENDOR::PACK@VERSION>",
            "program": /path/to/your/elf
}
```

If you are you using the `arm.device-manager` extension and the device is detected, you may be able to oinfer some of the details:

```jsonc
{
            "name": "Debug",
            "type": "embedded-debug",
            "request": "launch",
            "serialNumber": "${command:device-manager.getSerialNumber}",
            "cmsisPack": "${command:device-manager.getDevicePack}",
            "program": /path/to/your/elf
}
```

**Comment by daldr-ntml** [comment]
Thanks for your answer. Do I need any other specific extensions besides ‘Arm Embedded Debugger’ for debugging?

Do I need OpenOCD and / or gdb?

I guess your extension offers similar functionality to the 'Cortex-Debug' extension?

**Comment by thegecko** [comment]
> Do I need any other specific extensions besides ‘Arm Embedded Debugger’ for debugging?

No, although the extension is included with our other tooling in the [Keil Studio Extension Pack](https://marketplace.visualstudio.com/items?itemName=Arm.keil-studio-pack)

> Do I need OpenOCD and / or gdb?

No, a basic step debugger is included in the extension

> I guess your extension offers similar functionality to the 'Cortex-Debug' extension?

Our pack includes the SVD viewer originally developed as part of Cortex Debug, but doesn't yet include memory viewing.

The biggest difference however is that this debugger works natively in a browser environment (e.g. vscode.dev or GitHub codespaces).

**Comment by daldr-ntml** [comment]
Thanks for your answer. Please will you tell me what is the appropriate way to create launch.json in this context?

**Comment by thegecko** [comment]
> Please will you tell me what is the appropriate way to create launch.json in this context?

In which context?

**Comment by daldr-ntml** [comment]
As I understand it, if using the VS Code debugger, there is a 'create a launch.json file link' in the Run start view. (Ref. [Launch configurations](https://code.visualstudio.com/Docs/editor/debugging#_launch-configurations)).

How does one create launch.json  if using the ‘Arm Embedded Debugger’ extension? (That's what I meant by this context).

**Comment by thegecko** [comment]
It's specified as outlined in my post above. The debugger used is set by the type:

```
{
      ...
      "type": "embedded-debug",
      ...
}
```

**Comment by daldr-ntml** [comment]
Thanks.

**Comment by daldr-ntml** [comment]
@thegecko I would be grateful for some more guidance please. I really don't understand the workflow required for what I want to do.

I want to develop Cortex-M4 code on an Arm MPS2 dev. board.  I have chosen to use the **startup_Cortex-M4_AC6** project, supplied with Arm Development Studio, as an example to get started.

One way of running code on the MPS2 is to copy the elf file to the board's flash drive (which appears as a USB drive in Windows) and to specify that elf file in images.txt. When rebooted, the elf file should load and execute, but in my case it doesn't. I guess this is not the right place to discuss that problem, but it leads me on to trying to load the elf using the Arm Embedded Debugger extension instead.

I want to connect to the MPS2 using the CMSIS-DAP USB interface, so I assume I should create a **embedded-debug.daplink-flash:Flash Device (DAPLink)** task. Is that correct?

Here is my task:

    {
    	"version": "2.0.0",
    	"tasks": [
    		{
    			"type": "embedded-debug.daplink-flash",
    			"serialNumber": "${command:device-manager.getSerialNumber}",
    			"program": "/Users/daldrich/Development Studio Workspace/startup_Cortex-M4_AC6/startup_Cortex-M4_AC6.axf",
    			"problemMatcher": [],
    			"label": "embedded-debug.daplink-flash: Flash Device (DAPLink)"
    		}
    	]
    }

When I run that task will it program the elf into Flash on the MPS2?

Will that overwrite the existing boot loader, or whatever is there, that loads an elf from the Flash drive?

In other words, is this the correct approach?

**Comment by thegecko** [comment]
If your board supports DAPLink (and is up to date), that configuration should flash it. 

Although DAPLink usually expects binaries without symbols (hex or bin).

**Comment by daldr-ntml** [comment]
Thanks. Does the CMSIS-DAP USB interface use DAPLink?

**Comment by thegecko** [comment]
CMSIS-DAP is a protocol, DAPLink is firmware which supports CMSIS-DAP.

DAPLink generally supports drag-n-drop flashing (Mass storage device), CMSIS-DAP compliant targets don't have to.

**Comment by thegecko** [comment]
@DavidA2014 Were you able to get this working?

**Comment by daldr-ntml** [comment]
I didn't make any progress with this. Please feel free to close this issue.
